### PR TITLE
Fix code coverage in specific path mode

### LINKED ISSUE #44593: Requesting code coverage for a specific path misses lines
The option to track code coverage for files in a given path (https://github.com/JuliaLang/julia/pull/44359) appears to miss a significant number of lines compared to `--code-coverage=user`.

As seen in this Octavian CI run on julia nightly
https://github.com/JuliaLinearAlgebra/Octavian.jl/pull/143

Screen Shot 2022-03-12 at 6 31 44 PM

Screen Shot 2022-03-12 at 6 36 37 PM

### PR DESCRIPTION
**Pull request by IanButterworth**
Fixes https://github.com/JuliaLang/julia/issues/44593

### CODE REVIEW

File: base/compiler/utilities.jl [hunk 1]
```diff
@@ -359,13 +359,17 @@ inlining_enabled() = (JLOptions().can_inline == 1)
...
         m = moduleroot(m)
         m === Core && return false
         isdefined(Main, :Base) && m === Main.Base && return false
         return true
-    elseif cov == 2
+    elseif cov == 2 # all
         return true
+    elseif cov == 3 # path-specific
+        m = moduleroot(m)
+        mpath = pathof(m)
```
[the remark this answers is not in the document]
Comment (IanButterworth) [reply base/compiler/utilities.jl]: I've tried to make coverage info emission file-based but I'm hitting a BoundsError, so it appears the index handling isn't right

[the remark this answers is not in the document]
Comment (IanButterworth) [reply base/compiler/utilities.jl]: I couldn't figure out doing it on a line by line basis, so now it's enabled for all lines in a `CodeInfo` if any of the lines fall in tracked files. Tests pass locally.

**Comment by IanButterworth** [comment]
This fixes the codecov tests, and running the Octavian tests locally with CI enabled take the same time as master, so seems to not have a noticeable performance hit.

File: base/options.jl [hunk 2]
```diff
@@ -89,3 +89,8 @@ function unsafe_load_commands(v::Ptr{Ptr{UInt8}})
     end
     return cmds
 end
+
+function is_file_tracked(file::Symbol)
+    JLOptions().code_coverage == 3 || return false
```
Comment (IanButterworth) [inline base/options.jl:94]:
```suggestion
```

**Comment by IanButterworth** [comment]
@vtjnash I'm planning to merge this tomorrow, just in case you have any review/objection before then

File: base/compiler/optimize.jl [hunk 3]
```diff
@@ -558,12 +558,21 @@ function convert_to_ircode(ci::CodeInfo, sv::OptimizationState)
...
     changemap = fill(0, length(code))
-    labelmap = coverage ? fill(0, length(code)) : changemap
     prevloc = zero(eltype(ci.codelocs))
     stmtinfo = sv.stmt_info
     codelocs = ci.codelocs
     ssavaluetypes = ci.ssavaluetypes::Vector{Any}
     ssaflags = ci.ssaflags
+    if !coverage && JLOptions().code_coverage == 3
+        for codeloc in codelocs
+            if codeloc != 0 && is_file_tracked(ci.linetable[codeloc].file)
```
Comment (vtjnash) [inline base/compiler/optimize.jl:568]: You can directly scan the linetable (you don't need the iteration over codeloc to index it)

**Review by vtjnash — APPROVED** [review]
Lgtm. Some small comments

**Comment by IanButterworth** [comment]
Linux32 and win32 failures appear unrelated. 

The buildkite hang seems like a glitch? @DilumAluthge

**Pull request merged**